plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -375,9 +375,14 @@ private Throwable handleError(CpsBodySubContext sc, Throwable t) {
                 localThread = thread;
             }
             try {
-                var handler = localThread.getContextVariables().get(FailureHandler.class, localThread::getExecution, sc::getNode);
-                if (handler != null) {
-                    return handler.handle(sc, t);
+                var contextVars = localThread.getContextVariables();
+                // CpsFlowExecution.start sets the context variables for the initial CpsThread to null, so context
+                // variables for all subsequent threads remain null until a block step adds a variable to its body.
```

Comment:
At first, I thought the thread was somehow getting cleaned up prior to `CpsBodyExecution$FailureHandler` running due to some weird case with `WorkflowRun.doTerm` or `WorkflowRun.doKill`. After adding the log message added in `CpsThread.cleanUp` and turning up log levels, I realized that it was actually just that the context variables were always `null` for the thread in question.